src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,255 @@
+                                                      accessSecret(accessSecret),
+                                                      arn(arn), useSSL(useSsl) {}
+
+    bool operator==(const client_id_t &other) const {
+      return accessKey == other.accessKey &&
+             accessSecret == other.accessSecret &&
+             arn == other.arn && useSSL == other.useSSL;
+    }
+
+    struct hasher {
+      std::size_t operator()(const client_id_t &k) const {
+        return ((std::hash<std::string>()(k.accessKey)
+                 ^ (std::hash<std::string>()(k.accessSecret) << 1)) >> 1)
+               ^ (std::hash<std::string>()(k.arn) << 1)
+               ^ (std::hash<bool>()(k.useSSL) << 1);
+      }
+    };
+
+
+  };
+
+  struct message_t {
+    const AwsClient client;
+    const std::string& resourceARN;
+    const std::string& payload;
+
+    message_t(const AwsClient client, const std::string &resourceARN, const std::string &payload) :
+            client(client),
+            resourceARN(resourceARN),
+            payload(payload) {};
+  };
+
+  typedef std::unordered_map<client_id_t, AwsClient, client_id_t::hasher> ClientList;
+
+
+  typedef boost::lockfree::queue<message_t *, boost::lockfree::fixed_sized<true>> MessageQueue;
+
+  class Manager {
+  private:
+    MessageQueue messageQueue;
+    ClientList clientList;
+    std::thread runner;
+    bool stopped;
+    mutable std::mutex clientsLock;
+    CephContext *cct;
+    const ceph::coarse_real_clock::duration idle_time;
+
+    void publish_internal(message_t *msg) {
+      const std::unique_ptr<message_t> msg_owner(msg);
+
+      std::visit([&msg, this](auto &&arg) { this->publish_util(arg, msg->resourceARN, msg->payload); }, msg->client);
+    }
+
+    void publish_util(Aws::SNS::SNSClient *client, string resourceArn, string payload) {
+      ldout(cct, 1) << "Publishing.." << dendl;
+      Aws::SNS::Model::PublishRequest req;
+      req.SetMessage(Aws::String(payload));
+      req.SetTopicArn(Aws::String(resourceArn));
+      Aws::Client::ClientConfiguration configuration;
+      configuration.verifySSL = true;
+      configuration.connectTimeoutMs = 30000;
+
+      auto result = client->Publish(req);
+      if (result.IsSuccess()) {
+        ldout(cct, 1) << "Published..." << dendl;
+      } else {
+        ldout(cct, 1) << result.GetError().GetMessage() << dendl;
+      }
+    }
+
+    void publish_util(Aws::Lambda::LambdaClient *client, const string &resourceArn, const string &payload) {
+      Aws::Lambda::Model::InvokeRequest invokeRequest;
+      Aws::Utils::ARN arn((Aws::String(resourceArn)));
+      invokeRequest.SetInvocationType(Aws::Lambda::Model::InvocationType::RequestResponse);
+      invokeRequest.SetLogType(Aws::Lambda::Model::LogType::Tail);
+      invokeRequest.SetFunctionName(arn.GetResource());
+      std::shared_ptr<Aws::IOStream> body = Aws::MakeShared<Aws::StringStream>("Simple Allocation Tag");
```

Comment:
what sharing happens here?

Reply:
no sharing. 
but this is needed for the aws =sdk API